src/Lecture2.hs

```diff
@@ -70,7 +75,13 @@ return the removed element.
 >>> removeAt 10 [1 .. 5]
 (Nothing,[1,2,3,4,5])
 -}
-removeAt = error "TODO"
+removeAt :: Int -> [a] -> (Maybe a, [a])
+removeAt n = if n < 0 then (,) Nothing else go 0 []
+  where go :: Int -> [a] -> [a] -> (Maybe a, [a])
+        go _ filtered [] = (Nothing, filtered)
+        go index filtered (x:xs)
+          | index == n = (Just x, filtered ++ xs)
+          | otherwise = go (index + 1) (filtered ++ [x]) xs
```

Comment:
It is more efficient to construct lists using the cons operator (`:`) directly without introducing intermediate values or adding lists with `(++)` whenever possible. Adding elements to the end of the list is rather slow 🐌